src/model.py

```diff
@@ -1734,3 +1844,320 @@ def get_confidence(expression_logits, batch, current_epoch):
+
+                    # reshape to match query/key shapes
+                    first_transmission = first_transmission.permute(1, 0).unsqueeze(-1).expand(key.shape)
+                    second_transmission = second_transmission.permute(1, 0).unsqueeze(-1).expand(query.shape)
+
+                    query = query * second_transmission
+                    key = key * first_transmission
+
+                relation_attn, weights = self.components["relations"][f"stack_{stack}"][relation](
+                    # expects shape: [seq, batch, cnn_dim]
+                    query=query,
+                    key=key,
+                    value=value,
+                    key_padding_mask=mask,
+                    need_weights=True
+                )
+
+                attn_inter = torch.cat((cnn_outputs[first_task].permute(0, 2, 1), relation_attn.permute(1,0,2)), dim=-1)
+                attn_logits = self.components[first_task]["attn_linear"](attn_inter) 
+
+                # store task inputs for next stack, considering multiple relations for first task
+                if (first_task == prev_first) or (first_task == "shared") or (prev_first is None):
+                    same_task.append(attn_inter)
+                else:
+                    same_task = [attn_inter]
+
+                task_inputs[first_task] = torch.stack(same_task).mean(dim=0)  
+                prev_first = first_task    
+
+                # stack outputs for this relation (averaged after all stacks complete)
+                if first_task == "shared":
+                    for task in self.subtasks:
+                        outputs[task].append(attn_logits)
+                else:
+                    outputs[first_task].append(attn_logits)
+
+        # in case no interactions desired
+        if stack_count == 0:
+            for task in self.subtasks:
+                task_output = task_inputs[task]
+                if "cnn" in self.components[task].keys():
+                    task_output = self.components[task]["cnn"](task_output)
+                outputs[task] = [self.components[task]["linear"](task_output.permute(0, 2, 1))]
+
+        final_output = {
+            task: torch.stack(outputs[task]).mean(dim=0).permute(0, 2, 1)
+            for task in self.subtasks
+        }
+
+        return final_output
+
+    @staticmethod
+    def get_prob(epoch_count, constant=5):
+        """
+        Borrowed directly from https://github.com/ruidan/IMN-E2E-ABSA
+        
+        Compute the probability of using gold opinion labels in opinion transmission
+        (To alleviate the problem of unreliable predictions of opinion labels sent from AE to AS,
+        in the early stage of training, we use gold labels as prediction with probability 
+        that depends on the number of current epoch)
+
+        """
+        prob = constant/(constant+torch.exp(torch.tensor(epoch_count).float()/constant))
+        return prob
```

Comment:
Percentage of how much influence true labels should have on attention inputs, according to current epoch count. Borrowed from IMN codebase, w/ the added flexibility to configure the constant used, via `warm_up_constant` parameter in json-configs.